src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs

```diff
@@ -95,8 +96,310 @@ private static unsafe void CopyImpl(Array sourceArray, int sourceIndex, Array de
+
+            MethodTable* pDestMT = destTH.AsMethodTable();
+            nuint destSize = pDestArrayMT->ComponentSize;
+            ref object? srcData = ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(Unsafe.As<object?[]>(sourceArray)), sourceIndex);
+            ref byte data = ref Unsafe.AddByteOffset(ref MemoryMarshal.GetArrayDataReference(destinationArray), (nuint)destinationIndex * destSize);
+
+            for (int i = 0; i < length; i++)
+            {
+                object? obj = Unsafe.Add(ref srcData, i);
+
+                // Now that we have retrieved the element, we are no longer subject to race
+                // conditions from another array mutator.
+
+                ref byte dest = ref Unsafe.AddByteOffset(ref data, (nuint)i * destSize);
+
+                if (pDestMT->IsNullable)
+                {
+                    CastHelpers.Unbox_Nullable(
+                        ref dest,
+                        pDestMT,
+                        obj);
+                }
+                else if (obj is null)
+                {
+                    ThrowHelper.ThrowInvalidCastException_DownCastArrayElement();
+                }
+                else if (pDestMT->ContainsGCPointers)
+                {
+                    Buffer.BulkMoveWithWriteBarrier(
+                        ref dest,
+                        ref CastHelpers.Unbox(pDestMT, obj),
+                        destSize);
+                }
+                else
+                {
+                    Buffer.Memmove(
+                        ref dest,
+                        ref CastHelpers.Unbox(pDestMT, obj),
+                        destSize);
+                }
+            }
+        }
+
+        // Will box each element in an array of value classes or primitives into an array of Objects.
+        private static unsafe void CopyImplBoxEachElement(Array sourceArray, int sourceIndex, Array destinationArray, int destinationIndex, int length)
+        {
+            MethodTable* pSrcArrayMT = RuntimeHelpers.GetMethodTable(sourceArray);
+            TypeHandle srcTH = pSrcArrayMT->GetArrayElementTypeHandle();
+
+            Debug.Assert(!srcTH.IsTypeDesc && srcTH.AsMethodTable()->IsValueType);
+            Debug.Assert(!RuntimeHelpers.GetMethodTable(destinationArray)->GetArrayElementTypeHandle().AsMethodTable()->IsValueType);
+
+            MethodTable* pSrcMT = srcTH.AsMethodTable();
+
+            nuint srcSize = pSrcArrayMT->ComponentSize;
+            ref byte data = ref Unsafe.AddByteOffset(ref MemoryMarshal.GetArrayDataReference(sourceArray), (nuint)sourceIndex * srcSize);
+            ref object? destData = ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(Unsafe.As<object?[]>(destinationArray)), destinationIndex);
+
+            for (int i = 0; i < length; i++)
+            {
+                object? obj = RuntimeHelpers.Box(pSrcMT, ref Unsafe.AddByteOffset(ref data, (nuint)i * srcSize));
+                Unsafe.Add(ref destData, i) = obj;
```

Comment:
Is `WriteBarrier` required to safely write a reference type?

Reply:
This works fine. The assignment is going to compile into a write barrier call.